Privacy Enhancements

nicolae-leonte-go opened the pull request:
Adds the "Privacy Enhancements" feature to quorum.

The feature can be enabled via the `privacyEnhancementsBlock` chain config entry in genesis.json (a value must be agreed at the network level and all parties must run `geth init` to configure the value).

A new privacyFlag parameter was added to the sendTransaction API. The possible values are:
* 0 - StandardPrivate
* 1 - PartyProtection
* 3 - StateValidation

If the parameter is missing the value is assumed to be 0 - StandardPrivate.

The sendTransaction APIs have been enhanced to perform transaction simulation before submitting the transaction to the transaction pool in order to extract the necessary metadata for PartyProtection and StateValidation. 

StandardPrivate - is the value identifying the private transactions quorum has handled in all it's previous releases. We've aimed to keep the StandardPrivate behavior backward compatible (regardless of whether privacy enhancements are enabled/disabled) - unless a StandardPrivate transaction/contract attempts to affect privacy enhanced contracts.

PartyProtection - during simulation we detect all the contracts that the transaction affects. We then attach the necessary proofs/metadata to the transaction payload so that the other parties are able to validate that the sending party has access to all the affected contracts.

StateValidation - includes all the PartyProtection checks. In addition participants for StateValidation transactions are checked against the participants for all of the affected contracts. After simulation an execution hash is calculated from the state of of the affected contracts and is attached to the private transaction payload (which is then verified by the receiving nodes after the transaction is minted).

nicolae-leonte-go committed:
add exec hash to constellation api

nicolae-leonte-go committed:
add OrigTX/OrigTXHash to account and Privacy Enhancements with affected contract transactions

nicolae-leonte-go committed:
Merge branch 'master' into feature/PrivacyEnhancementsWithAffectedTransactions

nicolae-leonte-go committed:
Merge branch 'feature/newPrivacyEnhancements' into feature/PrivacyEnhancementsWithAffectedTransactions

# Conflicts:
#	core/state_transition.go
#	internal/ethapi/api.go
#	private/constellation/constellation.go
#	private/constellation/node.go
#	private/private.go

nicolae-leonte-go committed:
add execution hash (with naive implementation - hash of concatenated created/affected contract addresses)

nicolae-leonte-go committed:
propagate the preemptive check evm error so that the private transaction flow stops on the initiating node

nicolae-leonte-go committed:
check if current node has access to contract and propagate receive errors

trung committed:
updated with master

trung committed:
Merge branch 'private/master' into private/feature/PrivacyEnhancementsWithAffectedTransactions

trung committed:
refactored private transaction related logic

trung committed:
support send raw private transaction and fixed issue with transaction not found returned by 404 from TxManager

apratt3377 committed:
add privacyMetadata interface

trung committed:
refactored to use type EncryptedPayloadHash instead of []byte
not output full private payload via log info

trung committed:
added unit tests

trung committed:
justified contract.Address() over contract.CodeAddr usage

trung committed:
calculate merkle root for affected contracts, some minor logging improvement

trung committed:
go fmt and removed logging

trung committed:
minor cleanup

apratt3377 committed:
add storage of privacy metadata in db

trung committed:
support switching between Constellation and Tessera clients

trung committed:
Tessera client using JSON

trung committed:
added PSV flag support and refactoring

apratt3377 committed:
handle privacyMetadata errors

apratt3377 committed:
upstream merge

apratt3377 committed:
add psv flag to quorum

trung committed:
added tests and minor refactoring

apratt3377 committed:
don't send merkle root to tessera if not psc

apratt3377 committed:
cleanup psv logic

trung committed:
added test cases
remove PSV flag in request/response with Tessera

apratt3377 committed:
upstream merge and add privacyMetadata to tx

trung committed:
Merge pull request #7 from QuorumEngineering/storePrivacyMetadata

add storage of privacy metadata in db

apratt3377 committed:
take snapshot for stateTransition, uncomment merkleRoot check, uncomment raw tx flow

apratt3377 committed:
move gob encode/decode to separate funcs

apratt3377 committed:
gob encode error handling

trung committed:
support receive raw API

trung committed:
Merge pull request #8 from QuorumEngineering/storePrivacyMetadata

Store privacy metadata

trung committed:
used receiveRaw and minor clean up

apratt3377 committed:
rework psv logic

apratt3377 committed:
upstream merge

apratt3377 committed:
fix psv logic

trung committed:
fixed private transaction unit test, add a new unit test for api.go

trung committed:
Merge branch 'feature/PrivacyEnhancementsWithAffectedTransactions' of github.com:QuorumEngineering/quorum-mirror into private/feature/PrivacyEnhancementsWithAffectedTransactions

apratt3377 committed:
fix non-party precheck

trung committed:
added PrivacyFlag support

trung committed:
added unit test for api.go

trung committed:
merged

apratt3377 committed:
first look at adding privacy flag

trung committed:
added test cases for api.go

apratt3377 committed:
upstream merge

apratt3377 committed:
tests with new privacy flags

trung committed:
must fail transaction receipt when expected hashes are not 100% matched

trung committed:
added test cases

trung committed:
don't log addresses as it won't give much value

apratt3377 committed:
db err not found shouldnt cause node to crash, signals non-party

apratt3377 committed:
Merge branch 'feature/PrivacyEnhancementsWithAffectedTransactions' of github.com:QuorumEngineering/quorum-mirror into feature/PrivacyEnhancementsWithAffectedTransactions

apratt3377 committed:
upstream merge

apratt3377 committed:
add message call tests

trung committed:
some optimization in evm.go

trung committed:
added web3ext API for getting contract privacy metadata

trung committed:
removed check which can result inconsistent account nonce

apratt3377 committed:
add isLegacyFlag method for easier privacy checks

apratt3377 committed:
Merge branch 'newPrivacyFlags' of github.com:QuorumEngineering/quorum-mirror into newPrivacyFlags

trung committed:
sending and receiving PrivacyFlag to/from tx manager

trung committed:
minor refactoring

apratt3377 committed:
privacyMetadata error handling for simulation

apratt3377 committed:
Merge branch 'newPrivacyFlags' of github.com:QuorumEngineering/quorum-mirror into newPrivacyFlags

apratt3377 committed:
remove extra logging

trung committed:
added caching support for raw transactions

apratt3377 committed:
allow simulation to set correct nonce with default config for EIP158

apratt3377 committed:
Merge branch 'newPrivacyFlags' of github.com:QuorumEngineering/quorum-mirror into newPrivacyFlags

apratt3377 committed:
remove unneccessary file and logging, rework privacyMetadata error handling

trung committed:
Merge pull request #9 from QuorumEngineering/newPrivacyFlags

New privacy flags

nicolae-leonte-go committed:
fix for legacy transactions affected contract checks

trung committed:
Merge pull request #10 from QuorumEngineering/feature/fixLegacyTransactionsAffectedContractChecks

fix for legacy transactions affected contract checks

trung committed:
minor refactoring

trung committed:
refactored PrivacyFlag to make it strong typed

trung committed:
fixed unit tests

trung committed:
updated PrivateFlagStateValidation to include PartyProtection by default

trung committed:
added unit tests for EVM changes

apratt3377 committed:
require privacy flag be sent on message calls

apratt3377 committed:
upstream merge

trung committed:
added unit tests for state_transition

apratt3377 committed:
Merge branch 'feature/PrivacyEnhancementsWithAffectedTransactions' of github.com:QuorumEngineering/quorum-mirror into feature/PrivacyEnhancementsWithAffectedTransactions

apratt3377 committed:
add more state transition tests

apratt3377 committed:
remove msg To check bc is caught in affected loop and using privacy flag sent on msg call

apratt3377 committed:
Merge pull request #11 from QuorumEngineering/privacyFlagMsgCalls

Privacy flag msg calls

apratt3377 committed:
add more api and stateTransition tests

apratt3377 committed:
ac hash check and tests with public contract

apratt3377 committed:
merge from upstream including 1.8.18 upgrade (#12)

trung committed:
Revert "merge from upstream including 1.8.18 upgrade (#12)"

This reverts commit 5ec3dec47c2f0d4f984a4de8ea5f01885ebd642e.

trung committed:
merged with upstream

apratt3377 committed:
fix typos

trung committed:
merged with master

nicolae-leonte-go committed:
rename PrivacyFlagLegacy to PrivacyFlagStandardPrivate

nicolae-leonte-go committed:
Merge branch 'feature/PrivacyEnhancementsWithAffectedTransactions' into feature/master-1.9.7-withPE

nicolae-leonte-go committed:
fix merge issues & store privacy metadata in a trie structure

nicolae-leonte-go committed:
fix re-add error check missed during merge

nicolae-leonte-go committed:
add privacy enhancements block to quorum config

nicolae-leonte-go committed:
do not store links to the empty trie root hash and add unit tests for privacy metadata linker

nicolae-leonte-go committed:
fix lint issues

nicolae-leonte-go committed:
update PR actions to use specific tessera and qat branches

nicolae-leonte-go committed:
Update pr.yml

nicolae-leonte-go committed:
Update pr.yml

nicolae-leonte-go committed:
Update pr.yml

nicolae-leonte-go committed:
Update pr.yml

nicolae-leonte-go committed:
Update pr.yml

nicolae-leonte-go committed:
Update pr.yml

nicolae-leonte-go committed:
Update pr.yml

nicolae-leonte-go committed:
Update pr.yml

nicolae-leonte-go committed:
Update pr.yml

nicolae-leonte-go committed:
Update pr.yml

nicolae-leonte-go committed:
Update pr.yml

nicolae-leonte-go committed:
do not replace empty tx data field with empty encrypted payload hash (64 zeros)

nicolae-leonte-go committed:
Add PrivateTransactionManager Features API and prevent quorum from progressing with privacy enhancements enabled if the privacy manager does not support it

nicolae-leonte-go committed:
fix lint issues

nicolae-leonte-go committed:
fix lint issues

nicolae-leonte-go committed:
add minor component to version and refactor version parsing logic

nicolae-leonte-go committed:
Update pr.yml

nicolae-leonte-go committed:
move quorum validation from geth to startNode and simplify the validation logic for privacy enhancements block

nicolae-leonte-go committed:
fix lint issues

nicolae-leonte-go committed:
refactor the transaction manager features API

nicolae-leonte-go committed:
use camelCase for consts/variables that are not exported

nicolae-leonte-go committed:
Merge pull request #111 from QuorumEngineering/feature/master-1.9.7-withPE-PTMFeatures

Add PTM Features

nicolae-leonte-go committed:
add network upgrade acceptance tests

nicolae-leonte-go committed:
add extra information/context to errors

nicolae-leonte-go committed:
add extra information/context to errors

nicolae-leonte-go committed:
remove engine.NotInUsePrivateTxManager and enhance notinuse.PrivateTransactionManager (as well as using it as a base type for multiple transaction manager stub implementations in tests)

nicolae-leonte-go committed:
Merge branch 'master' into feature/master-1.9.7-withPE

# Conflicts:
#	core/private_state_test.go

nicolae-leonte-go committed:
fix lint issues

nicolae-leonte-go committed:
state transition refactoring

nicolae-leonte-go committed:
update logic to use pending/uncommitted contract state for execution hash calculations

nicolae-leonte-go committed:
fix lint issues and add extra test for the scenario where the execution hash calculation fails

nicolae-leonte-go committed:
remove redundant privateMessageContext fields, add comments, change log severity from error to debug when encountering post validation issues

nicolae-leonte-go committed:
Merge pull request #120 from QuorumEngineering/feature/master-1.9.7-withPE-refactor

state transition refactoring

nicolae-leonte-go committed:
print the privacy enhancements warning during first geth init,
add errors for GetStatePrivacyMetadata,
replace log.Error with fmt.Errorf and reduce severity for a number of log statements,
add privacyEnhancementsBlock to ChainConfig.String()

nicolae-leonte-go committed:
add info message that privacy enhancements are configured to be enabled from the specified block height

nicolae-leonte-go committed:
fix lint (goimports) issues

vsmk98 committed:
internal/ethapi: validation to check if private contract is existing on the node initiating the contract. modified tests in line and some code clean up

vsmk98 committed:
internal/ethapi: refactored to align FillTransaction with SendTransaction call

vsmk98 committed:
internal/ethapi: added a condition to exclude private transaction processing for public transactions to checkAndHandlePrivateTransactions function.

vsmk98 committed:
Merge branch 'feature/master-1.9.7-withPE' into feature/psv-with-ce

# Conflicts:
#	cmd/geth/config.go
#	private/engine/notinuse/notInUsePrivateTxManager.go
#	private/private.go
#	private/privatetransactionmanager/node.go
#	private/privatetransactionmanager/tx_manager.go

Initial merge

nicolae-leonte-go committed:
switch from gob to rlp encoding for privacy metadata, rename privacy metadata context to handler, fix typos

vsmk98 committed:
Fixes to resolve build failures post merge. Fixed lynt issues and resolve unit test failures

nicolae-leonte-go committed:
Merge branch 'feature/master-1.9.7-withPE' into feature/psv-with-ce

nicolae-leonte-go committed:
finish renaming privateMessageContext to privateMessageHandler

nicolae-leonte-go committed:
Merge branch 'api-contract-validation' into feature/psv-with-ce

nicolae-leonte-go committed:
relax checks for StandardPrivate transactions during transaction submission

vsmk98 committed:
internal/ethapi: corrected typos. slight refactor of the code.

vsmk98 committed:
extension: changes to handle extension of PSV and PP contracts

nicolae-leonte-go committed:
return error when non SP transactions are received in state_transition and privacy enhancements are disabled

nicolae-leonte-go committed:
enable istanbul upgrade tests

vsmk98 committed:
Merge branch 'feature/psv-with-ce' into feature/CE-with-PSVPP

nicolae-leonte-go committed:
Merge branch 'master' into feature/psv-with-ce

# Conflicts:
#	common/types_test.go
#	core/state_transition.go
#	core/state_transition_test.go
#	ethclient/ethclient.go
#	ethclient/ethclient_test.go
#	ethclient/privateTransactionManagerClient.go
#	ethclient/privateTransactionManagerClient_test.go
#	extension/backend.go
#	extension/backend_test.go
#	extension/privacyExtension/state_setter.go
#	graphql/graphql.go
#	graphql/graphql_test.go
#	internal/ethapi/api.go
#	private/engine/notinuse/notInUsePrivateTxManager.go
#	private/engine/notinuse/notInUsePrivateTxManager_test.go
#	private/private.go
#	private/privatetransactionmanager/tx_manager.go

vsmk98 committed:
extension/privacyExtension: added unit test for setPrivacyMetaData. added error handling

nicolae-leonte-go committed:
fix merge issues

nicolae-leonte-go committed:
fix lint issues

vsmk98 committed:
Merge branch 'feature/psv-with-ce' into feature/CE-with-PSVPP

nicolae-leonte-go committed:
unit tests for FormatTerminalString

vsmk98 committed:
Merge pull request #127 from QuorumEngineering/feature/CE-with-PSVPP

extension: changes to handle extension of PSV and PP contracts

Krish1979 committed:
Update Overview.md

Update flow of contract extension with privacy enhanced contracts

Krish1979 committed:
Update New batch recovery process

Krish1979 committed:
Overview for Privacy Enhancements

Krish1979 committed:
Create Privacy Enhancement apis.md

Krish1979 committed:
Add files via upload

Krish1979 committed:
Update Privacy Enhancement links

nicolae-leonte-go committed:
remove stale TODOs and fix/update comments

nicolae-leonte-go committed:
Merge remote-tracking branch 'qe/feature/psv-with-ce' into feature/psv-with-ce

Krish1979 committed:
Update Overview.md

nicolae-leonte-go committed:
pe documentation rewording

nicolae-leonte-go committed:
Merge branch 'feature/psv-with-ce' into feature/psv-docs-fixes

# Conflicts:
#	docs/Privacy/Privacy-Enhancements/Overview.md

Krish1979 committed:
Merge pull request #132 from QuorumEngineering/feature/psv-docs-fixes

Feature/psv docs fixes

nicolae-leonte-go committed:
remove stale TODOs and pe documentation fixes

nicolae-leonte-go committed:
add simulation limitations and contract interaction matrix

nicolae-leonte-go committed:
update the contract interaction matrix diagram

nicolae-leonte-go committed:
update the contract interaction matrix diagram

vsmk98 committed:
extension: chnaged uuid generation logic to support recoverability
private: added encrypt and decrypt APIs to tessera

nicolae-leonte-go committed:
fix private transaction trace for privacy enhancements

vsmk98 committed:
Merge branch 'feature/psv-with-ce' into CE-recoverability

prd-fox committed:
Fixed issue with regex not picking up minor version correctly

prd-fox committed:
Merge branch 'feature/psv-with-ce' of github.com:QuorumEngineering/quorum into CE-recoverability

vsmk98 committed:
common: removed unnecessary code.
extension: corrected typo in code comments

vsmk98 committed:
Merge remote-tracking branch 'qeng/CE-recoverability' into CE-recoverability

nicolae-leonte-go committed:
relax ACOTH count check (to cover the situation where multiple affected cotnracts share the same ACOTH)

vsmk98 committed:
Merge pull request #143 from QuorumEngineering/CE-recoverability

PR for contract extension in tessera recover scenario handling

nicolae-leonte-go committed:
Merge branch 'master' into feature/psv-with-ce

# Conflicts:
#	docs/Privacy/Contract-Extension/Overview.md
#	docs/Privacy/Tessera/Tessera Services/Transaction Manager.md

nicolae-leonte-go committed:
gh actions change to pull and build tessera-pe docker image

nicolae-leonte-go committed:
Merge branch 'master' into feature/privacyEnhancements

# Conflicts:
#	core/state_prefetcher.go
#	core/state_processor.go
#	eth/api_tracer.go

nicolae-leonte-go committed:
fix lint issues

nicolae-leonte-go committed:
remove privacy enhancements documentation after moving it to the doc.goquorum repo

nicolae-leonte-go committed:
update pr gh actions to use locally built tessera and quorum-acceptance-tests docker images

nicolae-leonte-go committed:
update pr gh actions alignment

nicolae-leonte-go committed:
update pr gh actions alignment

nicolae-leonte-go committed:
update pr gh actions to include more acceptance tests

nicolae-leonte-go committed:
update pr gh actions to include more acceptance tests

nicolae-leonte-go committed:
update pr gh actions to include more acceptance tests

nicolae-leonte-go committed:
update pr gh actions to run tests sequentially

nicolae-leonte-go committed:
update the tessera version with privacy enhancements features

nicolae-leonte-go committed:
kick off pr gh actions

nicolae-leonte-go committed:
fix version unit tests

nicolae-leonte-go committed:
kick off pr gh actions

nicolae-leonte-go committed:
kick off pr gh actions

nicolae-leonte-go committed:
kick off pr gh actions

nicolae-leonte-go committed:
kick off pr gh actions

trung reviewed (changes requested):
**Tollgate Note:** `.github/workflows/pr.yml` changes are to enable comprehensive acceptance tests for this PR. Those tests will be included in `quorum-acceptance-tests` repo. Before merging, please rollback the changes.

jbhurat commented on core/rawdb/database_quorum.go:
```diff
@@ -49,10 +53,23 @@ func GetPrivateStateRoot(db ethdb.Database, blockRoot common.Hash) common.Hash {
+func WritePrivacyMetadataStateRootForPrivateStateRoot(db ethdb.KeyValueWriter, privateStateRoot, privacyMetadataRoot common.Hash) error {
+	err := db.Put(append(privateRootToPrivacyMetadataRootPrefix, privateStateRoot[:]...), privacyMetadataRoot[:])
+	if err != nil {
+		return fmt.Errorf("unable to persist mapping between private state root to privacy metadata root. Cause: %v", err)
```
We should not wrap the error with the message here, instead just return whatever `db.Put()` returns. The calling function should provide helpful log if required

nicolae-leonte-go replied:
fixed

nicolae-leonte-go committed:
Remove the error wrapping in WritePrivacyMetadataStateRootForPrivateStateRoot. The calling function provides sufficient context.

chris-j-h commented on mkdocs.yml:
```diff
@@ -56,6 +55,9 @@ nav:
                 - APIs: Privacy/Contract-Extension/ContractExtension apis.md
+            - Privacy Enhancements:
+                - Overview: Privacy/Privacy-Enhancements/Overview.md
+                - APIs: Privacy/Privacy-Enhancements/Privacy Enhancement apis.md
```
these changes aren't necessary as `docs/` has been deprecated

nicolae-leonte-go replied:
removed

nicolae-leonte-go committed:
Remove leftover privacy enhancements docs changes (from mkdocs.yml).
Update the privacy enhancements warning tessera version (shown during geth init).

nicolae-leonte-go committed:
Get docker container info in case of acceptance tests failures.

jbhurat commented on extension/privacyExtension/state_set_utilities.go:
```diff
@@ -28,10 +28,27 @@ func setState(privateState *state.StateDB, accounts map[string]extension.Account
+
+	ptmHash, err := common.Base64ToEncryptedPayloadHash(hash)
+	if err != nil {
+		log.Debug("setting privacy metadata failed", "err", err)
```
Shouldn't this be logged as `Error`

vsmk98 replied:
Agree Jitu . Will change to `Error`

nicolae-leonte-go replied:
fixed

nicolae-leonte-go committed:
kick off pr checks after tessera and acceptance tests updates

vsmk98 commented on cmd/evm/runner.go:
```diff
@@ -217,6 +217,7 @@ func runCmd(ctx *cli.Context) error {
 	execTime := time.Since(tstart)
 
 	if ctx.GlobalBool(DumpFlag.Name) {
+		//TODO - Quorum - what is the point of doing both (commit does IntermediateRoot anyway)?
```
Please get rid of the `TODO` here. I know the call is redundant, but I would leave it there, keeping it aligned with upstream. Also this should not have any impact processing time

nicolae-leonte-go replied:
removed

vsmk98 commented on core/state_transition_pmh.go:
```diff
@@ -0,0 +1,145 @@
+		//when privacyMetadata should have been recovered but wasnt (includes non-party)
+		//non party will only be caught here if sender provides privacyFlag
+		if err != nil && privacyFlag.IsNotStandardPrivate() {
+			return returnErrorFunc(nil, "PrivacyMetadata unable to be found", "err", err)
```
nitpick - do we want to change the message to `unable to find privacy metadata`?

nicolae-leonte-go replied:
error message updated

vsmk98 commented on core/state_transition_pmh.go:
```diff
@@ -0,0 +1,145 @@
+	actualACAddresses := pmh.stAPI.AffectedContracts()
+	log.Trace("Verify hashes of affected contracts", "expectedHashes", pmh.receivedPrivacyMetadata.ACHashes, "numberOfAffectedAddresses", len(actualACAddresses))
+	privacyFlag := pmh.receivedPrivacyMetadata.PrivacyFlag
+	actualACHashesLength := 0
```
`actualACHashesLength` can be removed as we are not using it after removing the validation on this.

nicolae-leonte-go replied:
removed

nicolae-leonte-go committed:
Merge branch 'master' into feature/privacyEnhancements

nicolae-leonte-go committed:
Remove clarified TODO and unused variables, reword error message.
Change log level from Debug to Error - when failing to convert an encrypted payload hash.

nicolae-leonte-go committed:
Add more docker info when a test fails

nicolae-leonte-go committed:
Merge branch 'master' into feature/privacyEnhancements

chris-j-h committed:
Remove redundant arg from handlePrivateTransaction method signature
Additionally add more version parsing tests

trung commented on core/rawdb/database_quorum.go:
```diff
@@ -27,9 +27,11 @@ import (
+	privateBloomPrefix                     = []byte("Pb")
+	quorumEIP155ActivatedPrefix            = []byte("quorum155active")
+	privateRootToPrivacyMetadataRootPrefix = []byte("PSR2PMDR")
+	emptyRoot                              = common.HexToHash("56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421")
```
it may be worth to comment on this value:
```
// emptyRoot is the known root hash of an empty trie. Duplicate from `trie/trie.go#emptyRoot`
```

nicolae-leonte-go replied:
added the suggested comment

trung commented on core/state/database.go:
```diff
@@ -109,16 +113,29 @@ func NewDatabase(db ethdb.Database) Database {
+	// Quorum - Privacy Enhancements
+	privacyMetadataLinker rawdb.PrivacyMetadataLinker
+	// End Quorum - Privacy Enhancements
 	codeSizeCache *lru.Cache
```
There are number of places having the practice of wrapping with `Quorum - xxx` and `End Quorum - xxx` which will result misleading documentation. In particular, `codeSizeCache` will have documentation as `End Quorum - Privacy Enhancements`.

It would be clearer to document the field and its intention in a single line. E.g.:
```go
type cachingDB struct {
	db *trie.Database
	// Quorum:  Privacy enhacements introducing privacyMetadataLinker which maintains mapping between private state root and privacy metadata root. As this struct is the backing store for state, this gives the reference to the linker when needed  
	privacyMetadataLinker rawdb.PrivacyMetadataLinker
```

It can be applied for new methods or  updating existing methods. Giving some context and motivation for the changes.

nicolae-leonte-go replied:
updated the comment you mentioned and a number of other similar comments.

trung commented on core/genesis.go:
```diff
@@ -248,9 +250,20 @@ func SetupGenesisBlockWithOverride(db ethdb.Database, genesis *Genesis, override
 		return newcfg, stored, compatErr
 	}
 	rawdb.WriteChainConfig(db, stored, newcfg)
+	if storedcfg.PrivacyEnhancementsBlock == nil && newcfg.PrivacyEnhancementsBlock != nil {
```
Can we remove `newcfg.PrivacyEnhancementsBlock != nil` condition?  
It seems already be included in the function `checkAndPrintPrivacyEnhancementsWarning()`

nicolae-leonte-go replied:
removed

nicolae-leonte-go committed:
Merge pull request #1 from chris-j-h/pe-simplify-handleprivatetransaction-method

Remove redundant arg from handlePrivateTransaction method signature

nicolae-leonte-go committed:
use the new /version/api to retrieve the tessera API version

nicolae-leonte-go committed:
add emptyRoot comment and remove unnecessary check for PrivacyEnhancementsBlock

nicolae-leonte-go committed:
remove // End Quorum comments that affect unrelated field information in structures and improve other comments

nicolae-leonte-go committed:
add extra docker info and container log files when tests fail

nicolae-leonte-go committed:
search for the latest version in the /version/api versions array

nicolae-leonte-go committed:
Merge branch 'master' into feature/privacyEnhancements

nicolae-leonte-go committed:
if an older version of tessera that is not capable of privacy enhancements is used invoke the octetstream version of the /sendsignedtx API

nicolae-leonte-go committed:
update /version/api response format

nicolae-leonte-go committed:
use tessera master branch

nicolae-leonte-go committed:
add tests with older tessera version (bypassing contract extension tests that are not supported by the older tessera)

nicolae-leonte-go committed:
get logs for all containers (even those that exited recently)

nicolae-leonte-go committed:
rollback pr.yml changes

trung reviewed (dismissed):
LGTM. Just make sure to rollback the `pr.yml` to make it clean as noted in my first comment.

vsmk98 reviewed (dismissed):
LGTM

nicolae-leonte-go committed:
Merge pull request #2 from nicolae-leonte-go/feature/rollbackPR.YML

Rollback pr.yml

trung reviewed (approved):
LGTM
